core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -62,6 +62,24 @@ public GatewayAllocator(Settings settings, final TransportNodesListGatewayStarte
         this.replicaShardAllocator = new InternalReplicaShardAllocator(settings, storeAction);
     }
 
+    /**
+     * Returns true if the given shard has an async fetch pending
+     */
+    public boolean hasFetchPending(ShardId shardId, boolean primary) {
+        if (primary) {
+            AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> fetch = asyncFetchStarted.get(shardId);
+            if (fetch != null) {
+                return fetch.getNumberOfInFlightFetches() > 0;
```

Comment:
So, the issue with that is that some shards are completely assigned no problem without the async fetch actually being non-null, so if I assume that a null fetch means that it has one pending, they would show as needing a pending fetch when in fact they're already assigned and completely fine.

I think the fetch pending should only return true if there's actually a fetch pending, rather than if there is no fetch at all